lib/Dialect/FIRRTL/Transforms/LowerDomains.cpp

```diff
@@ -0,0 +1,516 @@
+struct AssociationInfo {
+  /// The DistinctAttr (annotation) that is used to identify the port.
+  DistinctAttr distinctAttr;
+
+  /// The port's location.  This is used to generate exact information about
+  /// certain property ops created later.
+  Location loc;
+};
+
+/// Track information about the lowering of a domain port.
+struct DomainInfo {
+  /// An instance of an object which will be used to track an instance of the
+  /// domain-lowered class (which is the identity of the domain) and all its
+  /// associations.
+  ObjectOp op;
+
+  /// The index of the optional input port that will be hooked up to a field of
+  /// the ObjectOp.  This port is an instance of the domain-lowered class.  If
+  /// this is created due to an output domain port, then this is nullopt.
+  std::optional<unsigned> inputPort;
+
+  /// The index of the output port that the ObjectOp will be connected to.  This
+  /// port communicates back to the user information about the associations.
+  unsigned outputPort;
+
+  /// A vector of minimal association info that will be hooked up to the
+  /// associations of this ObjectOp.
+  SmallVector<AssociationInfo> associations{};
+};
+
+/// Struct of the two classes created from a domain, an input class (which is
+/// one-to-one with the domain) and an output class (which tracks the input
+/// class and any associations).
+struct Classes {
+  /// The domain-lowered class.
+  ClassOp input;
+
+  /// A class tracking an instance of the input class and a list of
+  /// associations.
+  ClassOp output;
+};
+
+/// Thread safe, lazy pool of constant attributes
+class Constants {
+
+public:
+  Constants(MLIRContext *context) : context(context) {}
+
+  // Return an empty ArrayAttr.
+  ArrayAttr getEmptyArrayAttr() {
+    if (!emptyArrayAttr) {
+      llvm::sys::SmartScopedLock<true> lock(mutex);
+      emptyArrayAttr = ArrayAttr::get(context, {});
+    }
+    return emptyArrayAttr;
+  }
+
+private:
+  /// Construct all the field info attributes.
+  void constructFieldNameAttrs() {
+    if (domainInfoIn)
+      return;
+    llvm::sys::SmartScopedLock<true> lock(mutex);
+    domainInfoIn = StringAttr::get(context, "domainInfo_in");
+    domainInfoOut = StringAttr::get(context, "domainInfo_out");
+    associationsIn = StringAttr::get(context, "associations_in");
+    associationsOut = StringAttr::get(context, "associations_out");
+  }
+
+public:
+  StringAttr getDomainInfoIn() {
+    constructFieldNameAttrs();
+    assert(domainInfoIn);
+    return domainInfoIn;
+  }
+
+  StringAttr getDomainInfoOut() {
+    constructFieldNameAttrs();
+    assert(domainInfoOut);
+    return domainInfoOut;
+  }
+
+  StringAttr getAssociationsIn() {
+    constructFieldNameAttrs();
+    assert(associationsIn);
+    return associationsIn;
+  }
+
+  StringAttr getAssociationsOut() {
+    constructFieldNameAttrs();
+    assert(associationsOut);
+    return associationsOut;
+  }
+
+private:
+  /// Mutex indicating that, when held, allows for mutation.
```

Comment:
mutex needs to be on read as well as write, but I think I hammered that enough in my earlier comment.

I am not categorically against relying on word-size writes being atomic, but language specification says this is a race and would prefer we resorted to that only in places where it really helps save us (and ideally only relied on it on platforms we know that's true for?).